File: TitleBar/js/js/scenario1-colors.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var titleBar = Windows.UI.ViewManagement.ApplicationView.getForCurrentView().titleBar;

    var page = WinJS.UI.Pages.define("/html/scenario1-colors.html", {
        ready: function (element, options) {

            useStandardColors.addEventListener("click", changeColors);
            useCustomColors.addEventListener("click", changeColors);
            transparentWhenExtended.addEventListener("click", changeColors);

            // Infer the radio button selection from the title bar colors.
            if (!titleBar.buttonBackgroundColor) {
                useStandardColors.checked = true;
            } else if (titleBar.buttonBackgroundColor.a > 0) {
                useCustomColors.checked = true;
            } else {
                transparentWhenExtended.checked = true;
            }
            reportColors();
        }
    });

    var customColors = {
        backgroundColor: { a: 255, r: 54, g: 60, b: 116 },
        foregroundColor: { a: 255, r: 232, g: 211, b: 162 },
        buttonBackgroundColor: { a: 255, r: 54, g: 60, b: 116 },
        buttonForegroundColor: { a: 255, r: 232, g: 211, b: 162 },
        buttonHoverBackgroundColor: { a: 255, r: 19, g: 21, b: 40 },
        buttonHoverForegroundColor: { a: 255, r: 255, g: 255, b: 255 },
        buttonPressedBackgroundColor: { a: 255, r: 232, g: 211, b: 162 },
        buttonPressedForegroundColor: { a: 255, r: 54, g: 60, b: 116 },
        inactiveBackgroundColor: { a: 255, r: 135, g: 141, b: 199 },
        inactiveForegroundColor: { a: 255, r: 232, g: 211, b: 162 },
        buttonInactiveBackgroundColor: { a: 255, r: 135, g: 141, b: 199 },
        buttonInactiveForegroundColor: { a: 255, r: 232, g: 211, b: 162 },
    };

    function changeColors() {
        if (useStandardColors.checked) {
            // Setting colors to null returns them to system defaults.
            titleBar.backgroundColor = null;
            titleBar.foregroundColor = null;
            titleBar.inactiveBackgroundColor = null;
            titleBar.inactiveForegroundColor = null;

            titleBar.buttonBackgroundColor = null;
            titleBar.buttonHoverBackgroundColor = null;
            titleBar.buttonPressedBackgroundColor = null;
            titleBar.buttonInactiveBackgroundColor = null;

            titleBar.buttonForegroundColor = null;
            titleBar.buttonHoverForegroundColor = null;
            titleBar.buttonPressedForegroundColor = null;
            titleBar.buttonInactiveForegroundColor = null;
        } else {
            // Title bar colors. Alpha must be 255.
            titleBar.backgroundColor = customColors.backgroundColor;
            titleBar.foregroundColor = customColors.foregroundColor;
            titleBar.inactiveBackgroundColor = customColors.inactiveBackgroundColor;
            titleBar.inactiveForegroundColor = customColors.inactiveForegroundColor;

            // Title bar button background colors. Alpha is respected when the view is extended
            // into the title bar (see scenario 2). Otherwise, Alpha is ignored and treated as if it were 255.
            var buttonAlpha = transparentWhenExtended.checked ? 0 : 255;
            customColors.buttonBackgroundColor.a = buttonAlpha;
            customColors.buttonHoverBackgroundColor.a = buttonAlpha;
            customColors.buttonPressedBackgroundColor.a = buttonAlpha;
            customColors.buttonInactiveBackgroundColor.a = buttonAlpha;

            titleBar.buttonBackgroundColor = customColors.buttonBackgroundColor;
            titleBar.buttonHoverBackgroundColor = customColors.buttonHoverBackgroundColor;
            titleBar.buttonPressedBackgroundColor = customColors.buttonPressedBackgroundColor;
            titleBar.buttonInactiveBackgroundColor = customColors.buttonInactiveBackgroundColor;

            // Title bar button foreground colors. Alpha must be 255.
            titleBar.buttonForegroundColor = customColors.buttonForegroundColor;
            titleBar.buttonHoverForegroundColor = customColors.buttonHoverForegroundColor;
            titleBar.buttonPressedForegroundColor = customColors.buttonPressedForegroundColor;
            titleBar.buttonInactiveForegroundColor = customColors.buttonInactiveForegroundColor;
        }

        reportColors();
    }

    function reportColors() {
        var result = "";
        for (var p in customColors) {
            var v = titleBar[p];
            result += "Custom " + p + " = " + JSON.stringify(v) + "\n";
        }
        currentColorsDiv.innerText = result;
    }
})();

File: Samples/TitleBar/cppwinrt/Scenario1_Colors.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario1_Colors.h"
#include "Scenario1_Colors.g.cpp"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::UI;
using namespace Windows::UI::ViewManagement;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Data;

namespace winrt::SDKTemplate::implementation
{
    Scenario1_Colors::Scenario1_Colors()
    {
        InitializeComponent();

        titleBar = ApplicationView::GetForCurrentView().TitleBar();

        // Infer the radio button selection from the title bar colors.
        if (titleBar.ButtonBackgroundColor() == nullptr)
        {
            UseStandardColors().IsChecked(true);
        }
        else if (titleBar.ButtonBackgroundColor().Value().A > 0)
        {
            UseCustomColors().IsChecked(true);
        }
        else
        {
            TransparentWhenExtended().IsChecked(true);
        }

    }

    void Scenario1_Colors::UseCustomColors_Click(IInspectable const&, RoutedEventArgs const&)
    {
        UpdateColors();
    }

    void Scenario1_Colors::UpdateColors()
    {
        if (UseStandardColors().IsChecked().Value())
        {
            // Setting colors to null returns them to system defaults.
            titleBar.BackgroundColor(nullptr);
            titleBar.ForegroundColor(nullptr);
            titleBar.InactiveBackgroundColor(nullptr);
            titleBar.InactiveForegroundColor(nullptr);

            titleBar.ButtonBackgroundColor(nullptr);
            titleBar.ButtonHoverBackgroundColor(nullptr);
            titleBar.ButtonPressedBackgroundColor(nullptr);
            titleBar.ButtonInactiveBackgroundColor(nullptr);

            titleBar.ButtonForegroundColor(nullptr);
            titleBar.ButtonHoverForegroundColor(nullptr);
            titleBar.ButtonPressedForegroundColor(nullptr);
            titleBar.ButtonInactiveForegroundColor(nullptr);
        }
        else
        {
            // Title bar colors. Alpha must be 255.
            titleBar.BackgroundColor(ColorHelper::FromArgb(255, 54, 60, 116));
            titleBar.ForegroundColor(ColorHelper::FromArgb(255, 232, 211, 162));
            titleBar.InactiveBackgroundColor(ColorHelper::FromArgb(255, 135, 141, 199));
            titleBar.InactiveForegroundColor(ColorHelper::FromArgb(255, 232, 211, 162));

            // Title bar button background colors. Alpha is respected when the view is extended
            // into the title bar (see scenario 2). Otherwise, Alpha is ignored and treated as if it were 255.
            uint8_t buttonAlpha = (uint8_t)(TransparentWhenExtended().IsChecked().Value() ? 0 : 255);
            titleBar.ButtonBackgroundColor(ColorHelper::FromArgb(buttonAlpha, 54, 60, 116));
            titleBar.ButtonHoverBackgroundColor(ColorHelper::FromArgb(buttonAlpha, 19, 21, 40));
            titleBar.ButtonPressedBackgroundColor(ColorHelper::FromArgb(buttonAlpha, 232, 211, 162));
            titleBar.ButtonInactiveBackgroundColor(ColorHelper::FromArgb(buttonAlpha, 135, 141, 199));

            // Title bar button foreground colors. Alpha must be 255.
            titleBar.ButtonForegroundColor(ColorHelper::FromArgb(255, 232, 211, 162));
            titleBar.ButtonHoverForegroundColor(ColorHelper::FromArgb(255, 255, 255, 255));
            titleBar.ButtonPressedForegroundColor(ColorHelper::FromArgb(255, 54, 60, 116));
            titleBar.ButtonInactiveForegroundColor(ColorHelper::FromArgb(255, 232, 211, 162));
        }

        m_propertyChanged(*this, PropertyChangedEventArgs(L"TitleBar"));
    }

    event_token Scenario1_Colors::PropertyChanged(PropertyChangedEventHandler const& handler)
    {
        return m_propertyChanged.add(handler);
    }

    void Scenario1_Colors::PropertyChanged(event_token const& token) noexcept
    {
        m_propertyChanged.remove(token);
    }
}

hstring winrt::to_hstring(Windows::Foundation::IReference<Windows::UI::Color> const& value)
{
    if (value == nullptr)
    {
        return L"null";
    }

    Color color = value.Value();

    return L"A: " + to_hstring(color.A) + L",  R: " + to_hstring(color.R) + L", G: " + to_hstring(color.G) + L", B: " + to_hstring(color.B);
}
